pkg/registry/core/service/storage/storage.go

```diff
@@ -234,3 +244,163 @@ func (r *GenericREST) defaultAServiceOnRead(service *api.Service) error {
 
 	return nil
 }
+
+func (r *GenericREST) beginCreate(obj runtime.Object, options *metav1.CreateOptions) (func(runtime.Object, error), error) {
+	service := obj.(*api.Service)
+
+	// try set ip families (for missing ip families)
+	// we do it here, since we want this to be visible
+	// even when dryRun == true
+	if err := r.tryDefaultValidateServiceClusterIPFields(service); err != nil {
+		return nil, err
+	}
+
+	// bag of clusterIPs allocated in the process of creation
+	// failed allocation will automatically trigger release
+	var toReleaseClusterIPs map[api.IPFamily]string
+
+	// node-port state to clean up
+	nodePortOp := portallocator.StartOperation(r.serviceNodePorts, dryrun.IsDryRun(options.DryRun))
+
+	// Our cleanup callback
+	finish := func(_ runtime.Object, result error) {
+		if result == nil {
+			el := nodePortOp.Commit()
+			if el != nil {
+				// these should be caught by an eventual reconciliation / restart
+				utilruntime.HandleError(fmt.Errorf("error(s) committing service node-ports changes: %v", el))
+			}
+		} else {
+			released, err := r.releaseClusterIPs(toReleaseClusterIPs)
+			if err != nil {
+				klog.Warningf("failed to release clusterIPs for failed new service:%v allocated:%v released:%v error:%v",
+					service.Name, toReleaseClusterIPs, released, err)
+			}
+		}
+		nodePortOp.Finish()
+	}
+
+	//
+	// Do allocations
```

Comment:
ACK.  To do it more transactionally we'd need a pretty different allocation system.  At the end of the day we have the repair loops which (try to) prevent long-term leaks, but there is very much a risk of allocating, commiting the allocation to etcd, panic, retry sorts of loops.  We run the repair controller at startup, which mitigates proper crashes, but soft-fails could still happen.

We could add etcd transactions... :)

Reply:
I'm most interested in getting the call order so that admission webhook and disabled alpha/beta field dropping happens *before* we compute the ports we need to allocate

If I'm reading this correctly, this currently gets called after admission but before the `PrepareFor{Create,Update}`... that seems likely to bite us if we do clever things in those methods in the future.